maptool: fix JS (extract application JARs from package)

rhendric opened the pull request:
Extract application JARs from package instead of using the uber-JAR release.

Packing the JARs together causes the JavaScript macro functionality not to work, as the scripting interface uses a ServiceLoader to find language providers via text file resources that share a common name. With one file per JAR, this works great. When multiple files with the same name are in one uber-JAR, all but one of them are shadowed, and the JavaScript language provider was getting lost.

## Description of changes

Resolves #246830.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2803

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2836

SuperSandro2000 commented on pkgs/games/maptool/default.nix:
```diff
@@ -99,15 +97,20 @@ stdenvNoCC.mkDerivation {
+      --add-flags "${lib.concatStringsSep " " jvmArgs} net.rptools.maptool.client.LaunchInstructions"
 
     runHook postInstall
   '';
```
Please get the url via src.url

rhendric replied:
I don't know what line this comment is meant to refer to, but please note that `src.url` points to a single release artifact, not the entire repository like it would if this used `fetchFromGitHub` or something.

SuperSandro2000 commented on pkgs/games/maptool/default.nix:
```diff
@@ -13,15 +13,10 @@
   pname = "maptool";
   version = "1.13.2";
-  repoBase = "https://github.com/RPTools/${pname}";
+  repoBase = "https://github.com/RPTools/maptool";
```
```suggestion
```

rhendric replied:
Nope, this deduplicates the base repository URL between the source and the update script.

SuperSandro2000 replied:
You can get this url via src.url. We do not need to cripple the URL.

rhendric replied:
Are you aware that the value of `src.url` is `https://github.com/RPTools/maptool/releases/download/1.13.2/maptool-1.13.2-x86_64.pkg.tar.zst`?

Are you proposing splitting on `/` and selecting the first n components, or stripping the last n? Because that seems way less transparent than defining a variable for the URL base. I don't see any problem with this approach.

rhendric replied:
I don't see how ‘You can get this url via src.url’ could possibly mean that, but... come on, is this a blocker? Does being the maintainer of this package grant me the leeway to take the oh-so-bold position that deduplication is good actually, because in the future I want it to be obvious if the source and the update script URLs drift apart?

Run a `grep -rF 'url = "${'` on Nixpkgs if you think I'm being some sort of maverick by using this pattern. It's not exactly uncommon. Criticisms of functionality are more than welcome, and violations of the Nixpkgs house coding style also make good feedback, but this is something else altogether and I beg one of you to just click the dang merge button on this months-old PR if this is the only nit left to pick.

wegank replied:
Definitely not a blocker for me. I'll give Sandro 24 hours to react before I merge it, so ping me next night if it's still not merged.

SuperSandro2000 commented on pkgs/games/maptool/default.nix:
```diff
@@ -13,15 +13,10 @@
-  icon = fetchurl {
-    url = "https://raw.githubusercontent.com/RPTools/${pname}/${version}/package/linux/MapTool.png";
-    hash = "sha256-xkVYjMprTanHu8r4b9PHORI8E1aJp+9KDSP5mqCE8ew=";
+    url = "${repoBase}/releases/download/${version}/maptool-${version}-x86_64.pkg.tar.zst";
```
```suggestion
    url = "https://github.com/RPTools/maptool/releases/download/${version}/maptool-${version}-x86_64.pkg.tar.zst";
```

SuperSandro2000 commented on pkgs/games/maptool/default.nix:
```diff
@@ -13,15 +13,10 @@
+    url = "${repoBase}/releases/download/${version}/maptool-${version}-x86_64.pkg.tar.zst";
+    hash = "sha256-Ntmro+t4qpP5BXW20t97ki0wt2NKaK5yQarsxDEKbb0=";
   };
 
```
We should set platform to x86_64

SuperSandro2000 commented on pkgs/games/maptool/default.nix:
```diff
@@ -37,13 +32,15 @@ let
 
   javafxModules = [ "base" "controls" "media" "swing" "web" "fxml" "graphics" ];
 
+  appClasspath = "share/${pname}";
```
We don't need to put this onto a variable

rhendric replied:
Any time a path is used more than once in a package, I think it should be a variable. The name of the variable documents the purpose of the path and gives editors an unambiguous way to find all the places this directory is referenced. This variable is used once when constructing the `classpath` variable, and once when copying the JARs to $out.

SuperSandro2000 commented on pkgs/games/maptool/default.nix:
```diff
@@ -76,6 +73,7 @@ stdenvNoCC.mkDerivation {
   inherit pname version src meta;
 
   dontUnpack = true;
+  dontConfigure = true;
   dontBuild = true;
   dontWrapGApps = true;
```
Does any of those do an unwanted change? Otherwise please keep them enabled.

rhendric replied:
`dontUnpack` fails the build if omitted, because the source file is detected as an archive but the default builder can't unpack it.

`dontWrapGApps` keeps an unnecessary wrapper layer from being added; when we make our own wrapper, we include the `gappsWrapperArgs` there.

`dontConfigure` and `dontBuild` keep the builder from printing
```
configuring
no configure script, doing nothing
building
no Makefile or custom buildPhase, doing nothing
```
which is just unhelpful noise. The construction of this package is such that it'd be nonsensical to hook or override the configure or build phases; there is no upside to including them.

SuperSandro2000 commented on pkgs/games/maptool/default.nix:
```diff
@@ -99,15 +97,20 @@ stdenvNoCC.mkDerivation {
-      "''${gappsWrapperArgs[@]}" \
-      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ ffmpeg ]} \
-      --add-flags '${lib.concatStringsSep " " jvmArgs} net.rptools.maptool.client.LaunchInstructions'
+    dest=$out/${appClasspath}
+    install -dm755 "$dest"
+    bsdtar -xf "$src" -C "$dest" --strip-components 4 opt/maptool/lib/app/{'*.jar',readme}
```
The code would be easier to understand if we didn't variablize dest

rhendric replied:
I'd agree if I didn't need to use `dest` twice, once to create it and once to copy things into it. Each time it's defined, it's used only in the following two lines; this really isn't much of a barrier to understanding.

SuperSandro2000 commented on pkgs/games/maptool/default.nix:
```diff
@@ -99,15 +97,20 @@ stdenvNoCC.mkDerivation {
+    install -dm755 "$dest"
+    makeWrapper ${jre}/bin/java "$dest"/${binName} \
+      "''${gappsWrapperArgs[@]}" \
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ ffmpeg ]} \
```
Is there an alternative to LD_LIBRARY_PATH?

rhendric replied:
This package uses hundreds of pre-compiled JARs, many of which contain binary libraries. The only alternative I can think of would be unpacking those JARs, patchelf-ing each library, and then reassembling the JARs, and I really think using `LD_LIBRARY_PATH` is less brittle than that.

rhendric committed:
maptool: extract application JARs from package

... instead of using the uber-JAR release.

Packing the JARs together causes the JavaScript macro functionality not
to work, as the scripting interface uses a ServiceLoader to find
language providers via text file resources that share a common name.
With one file per JAR, this works great. When multiple files with the
same name are in one uber-JAR, all but one of them are shadowed, and
the JavaScript language provider was getting lost.

wegank commented:
Result of `nixpkgs-review pr 262650` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
